### ISSUE TITLE
Qt popup menus sometimes disappear when navigating submenus

### ISSUE THREAD

**Issue by wmww** [MirServer/mir#1941]
To reproduce:
- Open Kate
- Open menu with lots of sub-menus (like `tools` at the top)
- Move up and down, opening submenus

Expected: 
- You can keep doing that as long as you like

Observed:
- After a couple submenus are opened and closed, the toplevel menu is closed and you have to open it again

Issue does not exist on Sway or Weston

**Comment by wmww** [comment]
On Sway and Weston I'm seeing a bunch of the following message:
> qt.qpa.wayland: setGrabPopup called with a parent, QtWaylandClient::QWaylandXdgSurface(0x55afb03ce4f0) which does not match the current topmost grabbing popup, QtWaylandClient::QWaylandXdgSurface(0x55afb0609ad0) According to the xdg-shell protocol, this is not allowed. The wayland QPA plugin is currently handling it by setting the parent to the topmost grabbing popup. Note, however, that this may cause positioning errors and popups closing unxpectedly because xdg-shell mandate that child popups close before parents

But interestingly, I'm not seeing that on Mir (where unexpected popup closing is actually happening). Not sure what that means.

**Comment by wmww** [comment]
Reading through the wayland-debug logs of kate on both Mir and sway, I've noticed some things:
- The above warning seems to be the first relevant divergence between the two. It's unclear what Sway and Mir are doing differently to cause it to only appear on Sway
- Later, it seems that Sway is deleting old popups when a new popup is created and Mir is not
- On Mir Kate is creating a toplevel for one of the later submenus instead of a popup. This causes the root popup to close, which presumably causes Qt to close the submenu as well, resulting in the vanishing popups observed.

**Comment by wmww** [comment]
I'm fairly convinced the warning has something to do with the bug. Testing with KDevelop, it seems that both the warning and bug are intermittent, with the popups generally vanishing when the warning is *not* shown.

**Comment by wmww** [comment]
Figured it out. Kate is using the current keyboard-focused window as the popup parent. On other compositors that's the toplevel window. This isn't, but Qt notices this and fixes it (thus the warning). On Mir we're `wl_keyboard.enter`ing the popups. This means when switching from one popup to another (like moving down a list of sub-menus), it tries to use the previous popup as the new one's parent. The previous popup's shell surface has already been destroyed, so this fails in a way Qt can't fix. It instead tries to make a new toplevel. This causes the parent popup to become unmaped, which Qt notices and responds to by destroying the child popup it just created.

TLDR: fix is to not give popups keyboard focus
